### PR TITLE
add roadmap to docs

### PR DESCRIPTION
**Pull request by jhamman**
cc @rabernat @tomaugspurger @jbednar

### CODE REVIEW

File: doc/roadmap.rst [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+`{'time': 10}`) but this may evolve in the future. As we describe below,
+additional utilities for shuffling, sampling, and caching may provide enhanced
+batch generation functionality
+
+Shuffle and Sampling APIs
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+When training machine-learning models in batches, it is often necessary to
+selectively or randomly sample from your training data. Xbatcher can help
+facilitate seemless shuffling and sampling by providing APIs that opperate on
```
Comment (jbednar) [inline doc/roadmap.rst]:
```suggestion
facilitate seamless shuffling and sampling by providing APIs that operate on
```

**Comment by jhamman** [comment]
> Looks good. It would be nice to clarify with some more specific examples of the sorts of batching that are envisioned (spatial stencils)?

+1, @rabernat can you jot down some stencil language and I'll include it here.

**Comment by rabernat** [comment]
A "stencil" refers to a sub-region of a larger array that is used to generate a single "sample" for a ML workflow. For example, imagine we have a massive, high-resolution dataset of temperature over the whole earth of dimensions (1800 x 3600). We want to train a model to predict something at a specific point, based on a local neighborhood / radius of influence, using this data as inputs. But we know, for example, that our prediction in South America should not depend on the values in Northern Asia. So we define a "stencil" (e.g. 10 x 10) which is used to generate samples by iterating over the whole domain. These samples could be overlapping or non-overlapping. Xbatcher can help us with this by reshaping the input data as needed to generate these samples.

Note that this is somewhat overlapping with the concept of filters in a CNN. However, the difference here is that we assume that the region defined by a stencil is essentially independent from the rest of the domain, and the the machine learning problem is localized within the stencil.

**Comment by jbednar** [comment]
Stencils are also used in non-ML image-processing contexts, for any algorithms that work locally on regions of the image. Is supporting those sorts of algorithms outside the scope of xbatcher?  Fine if so; just curious...

**Comment by rabernat** [comment]
xbatcher basically just creates generators to iterate through ND-arrays in different ways. What you do with these generators is up to you! So definitely not out of scope.

**Comment by jhamman** [comment]
If it is alright with you two, I propose we merge this as is and follow up with additions to the docs/roadmap with examples.

**Pull request merged**